### ISSUE TITLE
Why our flavor of WKB?

### ISSUE THREAD

**Issue by jyutzler** [opengeospatial/geopackage#216]
I was recently asked the following question:

> Why does the geometry field in GeoPackage not use the same flavor of WKB/EWKB that is used in PostreSQL?

I was only tangentially involved in the project when this issue was being discussed and do not recall the specifics. If someone can answer this properly, please let us know so that I can reply. I will also add the response to the FAQ.

**Comment by jyutzler** [comment]
For the related discussion, see this tweet and its continuations: https://twitter.com/mafonso333/status/729080576054050816

**Comment by pepijnve** [comment]
- EWKB is not an OGC or ISO standard; it's an ad-hoc PostGIS extension. Didn't seem like a good idea to standardise it in GeoPackage
- Spatialite also uses an ad hoc extension/modification to WKB. Same problem as EWKB
- OGC/ISO WKB does not contain all the information we wanted to store (envelope + SRID)
- Prefixing OGC/ISO WKB rather than modifying it seemed like the cleanest solution. It allows standard WKB parsers/marshallers to be used for the bulk of the geometry. Extra information is in the header.

So in a nutshell the SWG opted for (header + standard wkb) rather than (modified wkb). I lobbied pretty hard for this so if anyone's to blame I guess it's me :)

**Comment by jyutzler** [comment]
Closed by #218

**Comment by taviroquai** [comment]
Hello,

Please open this issue :)

Just would like to leave here some ideas about dealing with a non-standard WKB. It is important to consider the universe of programming languages, the universe of geospatial software, direct and indirect implementation/operation costs and other real impacts missing here.

**Interoperability**
Operations like import/export from/to GeoPackage always will imply an additional parse header implementation.  
- Considering the universe of programming languages, at least, the most used 50 programming languages[1]. From the list of 50, only a limited number of it, can deal with low level data (byte level) operations. It would be important to know exactly which ones.
- Considering the most top 10 used programming languages, at least, there must be 10 parsing header implementations. It would be important to know how many implementations exists so far.
- Considering the universe of geospatial software[2], at least the most used, it is important to know which ones currently support/implement the GeoPackage WKB, and, which organizations/companies/individuals are willing to support it in future.

**Time Performance**[3]
- A parsing header implementation adds processing cost. It would be important to test parsing header performance against a big data set, let say 100000 records, AND at least in the most used top 10 programming languages. Note: the usage of the envelop present in the header, also implies a first parse operation.
- On a write operation, in most cases, implies an additional floating point calculation cost.

**Maintainability**[3]
 - Whenever the geometry changes, it implies an update/write operation of the header. It is important to consider the operation on a big data set ie. 100000 records.

Note: It is not easy to find good references about the problems exposed here, but these problems real and very frequent on software development area.
I also did not found any information related to the problems exposed here. It would be good to hear from the companies/software supporting the current GeoPackage WKB specification, how they deal with these problems and how they take advantage of the GeoPackage header specification.

[1] http://www.tiobe.com/tiobe_index
[2] https://en.wikipedia.org/wiki/List_of_spatial_analysis_software
[3] http://www.spinellis.gr/codequality/

**Comment by jyutzler** [comment]
@taviroquai 
I do not see any way to satisfy you here. It is not possible for this standard to satisfy every possible need. Our sweet spot includes the space from Shapefiles (too big or requiring things that are out of scope for GeoJSON) to smaller spatial databases (where PostGIS is overkill). If you find that using GeoPackage is causing your architecture to tilt, maybe GeoPackage is not the right solution for you. 

Please keep in mind that other concerns took priority when this issue was hashed out by the SWG in 2012-3. There was a long public review period before the standard was adopted by OGC in 2014. Even if the SWG decided that our approach was wrong, we would be unlikely to revisit the issue because changing the approach would break reverse compatibility. We're just going to have to live with it.

In addition, your request for scalability research is completely unreasonable unless you have a funding source we don't know about. We did conduct interoperability experiments as part of OGC Testbeds but with the limited funding available, we had to be selective about what was tested. If you find this unacceptable, my only suggestion is to get involved in OGC directly and change things from the inside.

**Comment by pepijnve** [comment]
@taviroquai I would like to understand the underlying reason for these concerns. Are you experiencing any performance or interoperability problems in your code that caused you to raise these issues?

FWIW, I understand the issues you listed. They're all valid points, but as with all things it's a matter of making tradeoffs. For geopackage, a binary format that requires a native C library in order to access it, I don't think putting a binary header in front of an existing binary blob is that big of an issue. What would the alternative be?

- Drop the header and loose the SRID and envelope? That drops essential information (the srid) which is absolutely necessary and comes with an implicit performance cost when using the rtree. The rtree needs the envelope of the geometry in order to update itself. If it is not present in the geometry blob it needs to be computed on the fly and the cost of that is much greater than that of parsing the header.
- Use a text encoding like WKT or something JSON based? That's just wasteful. The encoding on disk will be larger in many cases and parsing and marshalling will be slower. Quite a few SQL functions on geometry allow random access to points. In WKB encoding you can skip to the relevant parts fairly quickly and only extract the relevant information since the offsets are predictable. In a text encoding everything is variable length so that's simply not possible and you always have to do a full parse of the geometry for every operation.
- Use EWKB or Spatialite's internal encoding? Those are not standardised either; you would need decoders for these in each language as well. EWKB does not contain the envelope (rtree performance cost) and Spatialite makes the envelope mandatory (significant space overhead when encoding point data sets).

**Comment by taviroquai** [comment]
Hello Jeff and Pepijn :)

To be honest, I was expecting that you could answer to each of my points :) I'm not creating a war here, just trying to understand what bases (and use cases) motivated for this standard.

@jyutzler 
> Please keep in mind that other concerns took priority when this issue was hashed out by the SWG in 2012-3. There was a long public review period before the standard was adopted by OGC in 2014

I just got to know this standard in the last 2 weeks :) That's good to know Jeff. Is there any documentation about that review period? I would like to read. Any of the points that I presented was discussed?

> Even if the SWG decided that our approach was wrong, we would be unlikely to revisit the issue because changing the approach would break reverse compatibility. We're just going to have to live with it.

I understand.

> In addition, your request for scalability research is completely unreasonable unless you have a funding source we don't know about.

Yes you are right, it requires a deep research, and I thought that the bases for any standard should pass through these research. I think it is very important.

> We did conduct interoperability experiments as part of OGC Testbeds[...]

Is there any documentation available about those tests? I would like to read them.

@pepijnve 

> I would like to understand the underlying reason for these concerns.[..]For geopackage, a binary format that requires a native C library in order to access it, I don't think putting a binary header in front of an existing binary blob is that big of an issue.

I think that these concerns are almost "mandatory" when we want to create a standard that can be used by a large scale of users and a wide range of software/programming languages. I just realized that this standard may only be used in very limited and specific software architectures using C, Java, Python and a few more... The 3 points that you pointed out I completely agree!

Maybe you work on a different reality. Here is mine and is not from another Universe! :)

As a software developer, is mainly develop web applications using (mostly) Postgis, MapServer, OpenLayers, PHP and JavaScript. When using Postgis, it handles all the geometry operations for me, so I never had to parse their EWKB!
My GIS co-workers deal (mostly) with Postgis, MapServer, Shapefiles, ArcGIS, QuantumGIS and GDAL and these tools handle all the geometries operations.

So in my reality I can resume in the following points:
1. GDAL supports GeoPackage
2. GeoPackage EWKB is incompatible with Postgis.
3. Mapserver *may* support GeoPackage thought GDAL.
4. OpenLayers does not support GeoPackage.
5. In PHP there are some WKB parsing libraries. Hey I just created a GeoPackage parser in PHP (it needs more work) you can see [here](https://github.com/taviroquai/mapigniter2/blob/master/app/GeoPackage.php).
6. In Javascript also exists a GeoPackage implementation in [here](https://github.com/ngageoint/geopackage-js) but it seems that depends on Java.
7. Does ArcGIS supports GeoPackage?
8. QuantumGIS *may* support GeoPackage through GDAL.

At compatibility level, between the most currently used geospatial software, the options are very limited. I hope in the future more options will arise.

Now you understand the points that I exposed before :)

**Comment by danielbarela** [comment]
@taviroquai regarding point 6: The GeoPackage implementation you are referring to has been refactored away from Java.  The JavaScript version currently has full read capability and writing is next on the roadmap.

**Comment by tchaddad** [comment]
For #3 I can confirm that GeoPackage works nicely with Mapserver.

**Comment by cholmes** [comment]
As for #4, knowing the OpenLayers developers I'd guess that GeoPackage
support in OpenLayers should come before too long, using the library from
#6.

For #7 ArcGIS does support it. They were very involved in the standard
creation, and were a key voice in deciding the WKB formatting - it was
something we did our best to get right. See
http://www.esri.com/esri-news/releases/14-2qtr/esri-supports-ogc-geopackage-encoding-standard
I think there could be more support in their full suite of tools, but it's
definitely supported in some of the main packages.

I have successfully opened Geopackages on QGIS, it works out of the box in
recent versions. On my mac I can drag and drop a gpkg file and it will work.

Imho I don't actually think it makes sense for Geopackage binary to be
compatible with PostGIS EWKB - GeoPackage is a transfer format, and ewkb is
an internal format. So a toGPKG sql operation could make sense to add.

**Comment by jyutzler** [comment]
@taviroquai 
Most of the technical discussions occurred during internal SWG meetings and were not published as formal documents. Meeting minutes are available to SWG members and go back to December of 2012. It is pretty dry reading but it also shows the amount of work that it took to get here - a solid year of weekly meetings to get to 1.0 and fairly regular meetings until the end of 2015 to get to 1.1. The Wiki (also available to SWG members only) tracks many of the technical discussions, particularly those before August of 2013 when we started migrating to GitHub. 

The Testbed engineering reports such as https://portal.opengeospatial.org/files/?artifact_id=64189 do not appear to directly address your areas of concern. (Admittedly OGC does a poor job of making relevant engineering reports and discussion papers easy to find.)

I know it isn't easy to jump into this project. It is complex and some of the design decisions may seem counterintuitive at first glance. I can assure you that a lot of thought (and politicking) went into them. As SWG chairman I'm looking for ways to make it easier. This thread points out some areas where we could do a better job publicizing what we have done and for that I thank you.

**Comment by taviroquai** [comment]
Thank you Jeff. I will read the Testbed and give my feedback. Keep up the good working! :)

**Comment by taviroquai** [comment]
Hi, I enjoyed very much reading the Testbed use case. It would be good to realize the same use case but using GeoServer for server-side WPS/WFS processing and OpenLayers as web client and see how the results would be.

Also a use case that considers scalability would be good, ie. importing/exporting 100000 records between GeoPackage and Postgis.

**Comment by benstadin** [comment]
Hi,

I know I'm a little late to the discussion. I'm about to convert our proprietary format used for our 3D map rendering engine into something interoperable. So that the data can be used to only by the renderer, and worked on and visualized with different tools.
 
Thus I considered to base on GeoPackage, though. I need to define an extension for our projection-independent vector data tile format ("Heidelberg Sinusoidal Grid", HSG). For interoperability, I want to export a normal geopackage geometry table along the cells table (we call our tiles "cells" for a reason). In other words, the "cells" / tile tables are used for rendering while the normal geometry table is used for interoperability. But I have some other doubts now that I digg deeper into the spec. 

Some areas which I think need be improved, also I realize it would break compatibility (thus probably unlikely it will ever be changed):

**1. GeoPackage binary header**

I understand the rationale to have some additonal geometry info, e.g. for indexing. But the GeoPackageBinaryHeader as it is is wasting lot's of space. Consider lots of of location coordinates stored as simple points, such as we get from analytics data where we analyze user flows. The GeoPackageBinaryHeader as it is now seems a pain to me for several reasons:

a) It wastes disk space: The "magic", "version", and "flags", and "srs_id" info should be stored once in the gpkg_geometry_columns table as info rather than for each object. 

b) The srs_id info in the header is redundant. It is already defined in gpkg_contents. 

c) Geopackage claims to base itself on available standards. Yet, the envelope which could equally be defined as a standard WKB or else is defines as an own array of coordinates, with another redundant info about endianess and own defintion of the envelope with flags what could equally be a WKB polygon. 

d) Since the srs_id is defined in gpkg_contents it should apply to all geometries of the specified table. Again, having the srs_id in the header is redundant, and furthermore introduces potential errors: For example, the srs_id has to be changed in three places (gpkg_contents, gpkg_geometry_columns and individual geometries) and the behavior for geometries having a different srs_id than defined in the columns of contents table seems undefined. These could be interpreted either as erroneous entries which don't adhere to the specified srs_id, or as entries which have been added lately but weren't yet projected to the new srs_id, or it could be they are more "specific" e.g. overwriting the "default" srs_id with another (e.g. for geometries in a nearby UTM zone where you want certain geometries enforce a srs_id other than the "default"). Whatever is right here - it's overcomplicating implementation and the spec. 

**2. Multiple srs_id definitions**

I've already described the srs_id issue above within the geometry scope. The srs_id is defined in three places in total: 

- gpkg_contents
- gpkg_geometry_columns
- and within the header of each object

As already described above in 1. it is not well designed this way. It's defined at multiple places, whereas the scope of validity of a srs_id defined should be the table scope. 

**Missing elevation model**

The specification says this about z values:

_"The unit of measure for optional elevation(Z) values is determined by the CRS of the geometry; it is as-defined by a 3D CRS, and undefined for a 2D CRS."_

The problem is that it in practice a combination of an elevation model and a SRS is required. There is EPSG 4979, which is the 3D version of EPSG 4326. But many real world 3D applications and data require you to choose an elevation model. 

For an example, in the OpenStreetMap database z values can bes stored using WGS84 longitide / latitude and wgs84 ellipsoidal height, which they say is the most accurate combination [1].

But as you can imagine, the vast majority of data you get your hand on when mapping for example 3D building models has no wgs84 height info. 

Still, it is perfectly valid to have data with z values for WGS84 or even UTM and for example elevation model either defined as "local" or "wgs84". For rendering indoor buildings we use for example usually a UTM zone with local elevation [2, jump to 1:36m in the video]. You need to decide on your own how to convert that, but you need to store z values and have a hint about what elevation model is with your data. 

** Proposal**

I think the chances are very low that my proposal will be considered, since it would break compatibility somewhat. Nevertheless:

P1) It seems everybody is ok these days that there is only one geometry column allowed per table. It Seems totally acceptable, and it seems to me also that having multiple geometry columns would create more complexity. Therefore: Merge the definition of gpkg_geometry_columns with gpkg_contens, and remove gpkg_geometry_columns.

P2) Let gpkg_contens be the only place where the srs_id to be used by a table is defined. Should be ok now, since we also stick to the one-geometry-per-table rule and got rid of gpkg_geometry_columns. Thus, remove srs_id from gpkg_geometry_columns as well as from the geometry header (actually, I'd like to remove the geometry header completely, but let this be a seperate proposal). Also, remove srs_id from gpkg_contents and add srs_ref_id instead as a reference to the defintions in gpkg_spatial_ref_sys. See next point for rationale.

P3) Add a column "elevation_model" to gpkg_spatial_ref_sys with possible values of, just for example, "wgs84" and "local", and define that if it is null it's the elevation mode used by the srs. Also add an srs_ref_id column AUTOINC primary key. Let the gpkg_contens table refer to this srs_ref_id, and remove the srs_id from gpkg_contens. This is required because as described earlier there must be a combination possible to store elevation_model together with srs_id. Also, this allows to use other than EPSG codes, and use user-defined projection strings per table instead of only refering to the PESG code. 

P4) Add the version info, as it is currently defined in GeoPackageBinaryHeader, to the gpkg_contens table. Also add a geometry_binary_type to gpkg_contens with the same definition as-is currently defined for GeoPackageBinary within the header. Also add a field envelope_column to gpkg_contens.

P5) Finally, we can get rid of the geometry wrapper StandardGeoPackageBinary: Let the geometry column of a feature table be a plain vanilla WKB binary (or more precisely the same type as define in geometry_binary_type of gpkg_contens). Add a new but optional column "envelope" to the feature table as a replacement for the envelope within the header - let this also be a normal WKB binary. The envelope column is optional; if defined it should be also defined in gpkg_contens.envelope_column.

P6) Write a small translation utility to convert between tables defined in the new "Geopackage v2" layout and the current 1.2 spec. 

I think this will be a much cleaner spec, and allow even more simple adoption of GeoPackage. For example Postgis does not yet have (or unknown to me) support to create a GeoPackage binary - and there will be no more need to. All that's required for an application to support GeoPackage "v2" is what is support for WKB or WKT. 

I understand that this proposal will likely never get through. But to be honest, some of the points outlined above are so painful to us that I'd rather create a fork of the spec, base our software on it, and write a translation tool between this format and the original spec. So to keep us close to GPKG and ease translation and interoperability, and hope a future draft will improve on several of the points I mentioned. 

As an additional info: The performance thing regarding rtree indexing is a non-issue. It's perfectly possible to use the envelope as I described above, and if really required to do something fancy just add auxilary data to the prepared statements in SQLite. Having a wrapper object is not required at all from an implementation perspective if done right. 

Best regards,
Ben

[1] http://wiki.openstreetmap.org/wiki/Altitude
[2] http://www.deep-map.com
